redeclared-assigned-name: Allow dummy variables instead of hard-coded underscore

The check for `redeclared-assigned-name` has an [exception hard-coded for `_`](https://github.com/PyCQA/pylint/blob/master/pylint/checkers/base.py#L1555). If [`_` can not be used](https://stackoverflow.com/a/2745537/12345551) for unpacking return values, this is inconvenient. The check should instead allow all values that match [the `dummy-variables-rgx` configuration](https://docs.pylint.org/en/1.6.0/features.html).

### Steps to reproduce
```python
def some_method():
    return 1, 2, 3

_unused, a, _unused = some_method()
```

Run `pylint --disable=all --enable=redeclared-assigned-name --dummy-variables-rgx='_unused' sample.py`

### Current behavior
```
************* Module sample
sample.py:4:0: W0128: Redeclared variable '_unused' in assignment (redeclared-assigned-name)
```

### Expected behavior
no errors. `_unused` is a valid dummy variable name.

### pylint --version output
```
pylint 2.4.4
astroid 2.3.3
Python 3.7.5 (default, Nov  7 2019, 10:50:52) 
[GCC 8.3.0]
```
but the hard coded exception for "_" is [still in master](https://github.com/PyCQA/pylint/blob/master/pylint/checkers/base.py#L1555), so I don't see why any newer version would behave differently.

--- a/CONTRIBUTORS.txt
+++ b/CONTRIBUTORS.txt
@@ -365,3 +365,5 @@ contributors:
     Fixed dict-keys-not-iterating false positive for inverse containment checks
 
 * Anubhav: contributor
+
+* Benny Müller: contributor

--- a/ChangeLog
+++ b/ChangeLog
@@ -7,6 +7,10 @@ What's New in Pylint 2.5.0?
 
 Release date: TBA
 
+* Add `dummy-variables-rgx` option for `_redeclared-assigned-name` check.
+
+  Close #3341
+
 * Add a check for asserts on string literals.
 
   Close #3284

--- a/pylint/checkers/base.py
+++ b/pylint/checkers/base.py
@@ -1553,6 +1553,10 @@ class BasicChecker(_BasicChecker):
                 )
 
     def _check_redeclared_assign_name(self, targets):
+        dummy_variables_rgx = lint_utils.get_global_option(
+            self, "dummy-variables-rgx", default=None
+        )
+
         for target in targets:
             if not isinstance(target, astroid.Tuple):
                 continue
@@ -1562,6 +1566,8 @@ class BasicChecker(_BasicChecker):
                 if isinstance(element, astroid.Tuple):
                     self._check_redeclared_assign_name([element])
                 elif isinstance(element, astroid.AssignName) and element.name != "_":
+                    if dummy_variables_rgx and dummy_variables_rgx.match(element.name):
+                        return
                     found_names.append(element.name)
 
             names = collections.Counter(found_names)

--- a/tests/functional/r/redeclared_assigned_name.py
+++ b/tests/functional/r/redeclared_assigned_name.py
@@ -7,3 +7,12 @@ for SECOND, SECOND in enumerate(range(5, 10)):  # [redeclared-assigned-name]
 
 for FIRST, (SECOND, FIRST, SECOND) in enumerate(range(5, 10)):  # [redeclared-assigned-name]
     print(SECOND)
+
+for DUMM, DUMM in enumerate(range(5, 10)):  # [redeclared-assigned-name]
+    print(DUMM)
+
+for DUMMY, DUMMY in enumerate(range(5, 10)):
+    print(DUMMY)
+
+for _, _ in enumerate(range(5, 10)):
+    print(_)

--- a/tests/functional/r/redeclared_assigned_name.rc
+++ b/tests/functional/r/redeclared_assigned_name.rc
@@ -1,0 +1,2 @@
+[BASIC]
+dummy-variables-rgx=DUMMY

--- a/tests/functional/r/redeclared_assigned_name.txt
+++ b/tests/functional/r/redeclared_assigned_name.txt
@@ -1,3 +1,4 @@
 redeclared-assigned-name:3::Redeclared variable 'FIRST' in assignment
 redeclared-assigned-name:5::Redeclared variable 'SECOND' in assignment
 redeclared-assigned-name:8::Redeclared variable 'SECOND' in assignment
+redeclared-assigned-name:11::Redeclared variable 'DUMM' in assignment

